model/src/main/kotlin/utils/SnippetFinding.kt

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2023 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.model.utils
+
+import org.ossreviewtoolkit.utils.spdx.SpdxExpression
+
+/**
+ * A class representing a snippet finding. A snippet finding is a code snippet of [artifact] and [version], with the
+ * following [licenses], matching the code being scanned.
+ * It is meant to be reviewed by an operator as it could be a false positive.
+ */
+data class SnippetFinding(
+    /**
+     * The author/vendor of the component the code snippet is commit from.
+     */
+    val author: String,
+    /**
+     * The artifact name of the component the code snippet is commit from.
+     */
+    val artifact: String,
```

Comment:
Instead of using author, artifact and version any found snippet should imo point to a [ORT Package])https://github.com/oss-review-toolkit/ort/blob/main/model/src/main/kotlin/Package.kt) within ORT model this is how SBOM handle snippets see for example https://spdx.github.io/spdx-spec/v2.3/snippet-information/.

Also why does SnippetFinding Class not have location: TextLocation property to point to file path, start, and end line of the snippet?

Reply:
Don't forget that `SnippetFinding` is an abstraction of snippets and it should be mappeable from the results of all snippet scanners supported by ORT. 
To build your ORT Package, you need information such as the identifier. ScanOSS gives your the purl so it could be doable, but FossID does no give you the type of the identifier.
You also need the VCS of the file the snippet is coming from but for ScanOSS, it is under the form  "https://osskb.org/api/file_contents/6ff2427335b985212c9b79dfa795799f" which is a link to their KB, not the original file.

> Also why does SnippetFinding Class not have location: TextLocation property to point to file path, start, and end line of the snippet?

As mentioned in the commit message, it will be added in the future. It's not trivial because, while ScanOSS provides this information, FossId returns it under the following form:
 ```
"highlighting": "{\"id\":\"410668b1f35f8b27ff9ce345998448b6\",\"local_coverage\":0.9754,\"local_highlight\":{\"blocks\":[{\"byte_range\":{\"begin\":0,\"end\":712},\"char_range\":{\"begin\":0,\"end\":712},\"id\":\"abdc2b929a1b84f24155c27b752944ab\"},{\"byte_range\":{\"begin\":1395,\"end\":28013},\"char_range\":{\"begin\":1395,\"end\":28013},\"id\":\"a11cb7ad7af8d915193131a92d514ed7\"}],\"encoding\":\"UTF-8\",\"id\":\"3673e848c2d349e2f054691c952b3f2f\",\"pfm_format\":2},\"local_size\":475,\"remote_coverage\":1,\"remote_highlight\":{\"blocks\":[{\"byte_range\":{\"begin\":0,\"end\":27333},\"char_range\":{\"begin\":0,\"end\":27333},\"id\":\"871d76314c0e746c1b33d63e6c05a909\"}],\"encoding\":\"UTF-8\",\"id\":\"410668b1f35f8b27ff9ce345998448b6\",\"pfm_format\":2},\"remote_size\":475}"
		}
```

They have an API to get the matching lines but it is sub-performing. We opened a ticket in the (private) tracker to have it addressed.

Reply:
I believe that having the file location (scanned provenance) of the snippet finding is the most essential bit, as it is needed to make sense out of the finding. There it should be included in the initial MVP.

Regarding the way how the upstream snippet is specified `author, artifact, version, licenses, score`. How did you arrive at this one? Which snippet scanner did you consider when designing this model for?

Reply:
> @nnobelis Thank you for your clarifications, I think we should not base ORT model on what a specific tool currently provides but what we see as model that works with the rest of ORT. IMO Snippetfinding is should be start and end line and reference to a package.

As answered to @fviernau, in fact are two line ranges are needed: `sourceLocation` and `snippetLocation`.

> I am saying this as I would like to be able to curate snippet findings to resolve to package and be able to generate SBOM with snippets info.

I understand your need but in our current project, we do not need snippet results in a SBOM: we have a special reporter plugin for that.
Regarding the  possibility to curate snippet findings with `.ort.yml` (what we call "assessment" at Bosch) we are going to propose in some future PRs an `.ort.yml` model extension and an new ORT Evaluator. I think your concern can be addressed then ?

> 
> What do you thinking if we say introduce a in ORT Model a packageReference to hold some of the info you get from the snippet tools which can then be curated into a regular ORT Package?

For this I have unfortunately not enough experience with ORT data model. I would like @sschuberth, @mnonnenmacher, @fviernau opinion on this.

Reply:
> I understand your need but in our current project, we do not need snippet results in a SBOM: we have a special reporter plugin for that.

@nnobelis to my understanding any `ScanSummary` is and (after the refactoring to store scan results by provenance also in ORT result) will be implicitly linked to at least one or multiple `Package`s. I guess this fullfills @tsteenbe request to have a link to a package.

Reply:
But I thought @tsteenbe wanted a package per snippet ?!

@tsteenbe : did you really meant a link to the package for which the snippets have been listed ?

Reply:
> I understand your need but in our current project, we do not need snippet results in a SBOM: we have a special reporter plugin for that.

@nnobelis I prefer not to implement snippets in ORT that's just fulfill Bosch current project scope/needs but build something where we can see a path forwards that enable curating snippets and generate SBOM with snippets. As said for a snippet is a A) path to source (`sourceLocation`?), start and end line (snippetLocation/TextLocation?) and any other provenance info related to the package is should referenced - all of these snippet properties must be curatable at some point as no tool is perfect.

Reply:
@fviernau and I had talked about this PR a bit as part of another meeting to align SnippetFinding with existing ORT model components such as LicenseFinding and CopyrightFinding we came up with the following for SnippetFinding. Further discussion is needed such as how to curate snippetFinding but hope the below give you a better understanding of thoughts/ideas.

```
SnippetFinding
- location: textLocation e.g. file path, start and end lines
- license: SpdxExpression e.g. license applicable to the license
- score: String optional/nullable score of a snippet finding
- UpstreamSnippet
  - location: textLocation e.g. file path, start and end lines
  - Provenance
    - vcsInfo
    - sourceArtifact
  - purls e.g. array of purls optional/nullable as not every code fragment is part of a release.
```

Please describe the business case for including author. Why do you need the author or are after copyright holder?

Reply:
> license: SpdxExpression e.g. license applicable to the license

This is strange that `license` does not belong to `UpstreamSnippet` as it is a part of it.
Additional, could their be multiple licenses per snippet ? At least ScanOSS seems to believe so:

```
"licenses": [
      {
        "name": "MIT",
        "patent_hints": "no",
        "copyleft": "no",
        "checklist_url": "https://www.osadl.org/fileadmin/checklists/unreflicenses/MIT.txt",
        "osadl_updated": "2023-03-26T02:11:00+00:00",
        "source": "component_declared",
        "url": "https://spdx.org/licenses/MIT.html"
      }
    ],
```

Regarding the provenance, I think you will run in a wall: ScanOSS for instance does not give any information about the source repository:
```
"source_hash": "bd4bff27f540f4f2c9de012acc4b48a3",
"file_url": "https://osskb.org/api/file_contents/6ff2427335b985212c9b79dfa795799f",
```

As for the purl, as said above FossID gives you the `author`, `artifact`, `version` but not package type, so you would have to build some heuristics to find it.

For the author, we didn't need it explicitly, it was just a way of identifying the snippet. But in the end, this author is in your purl.

Reply:
> This is strange that license does not belong to UpstreamSnippet as it is a part of it.

Yes, on 2nd thought that makes more sense - improved version below:

```
SnippetFinding
- location: textLocation - file path, start and end lines
- score: String - score of a snippet finding, optional/nullable
- detectedSnippet
  - location: textLocation - file path, start and end lines
  - vcs: VcsInfoCurationData - nullable as not every snippet scanner contains this info
  - sourceArtifact: RemoteArtifact -  nullable as not every snippet scanner contains this info
  - purls: [string] e.g.  array of PURLs pointing to upstream package that include snippet, optional/nullable as not every code fragment is part of a release.
  - license: SPDXExpression e.g. license the snippetScanner thinks is applicable to the snippet
  - kb-url:
```

Also thought about curating snippet a bit more, below proposed SnippetFindingCuration is based on [LicenseFindingCuration.kt](https://github.com/oss-review-toolkit/ort/blob/main/model/src/main/kotlin/config/LicenseFindingCuration.kt)
```
SnippetFindingCuration
- path: string - glob to match the file path of a snippet finding.
- startLines: List<Int> - A matcher for the start line of a snippet finding.
- lineCount: Int - A matcher for the line count of a snippet finding
- concludedLicense: SpdxExpression - the concluded license as SPDX expression or [org.ossreviewtoolkit.utils.spdx.SpdxConstants.NONE] for no license,
- detectedLicense: SpdxExpression - a matcher for the detected license of a snippet finding
- reason: SnippetFindingCurationReason - the reason why the curation was made, out of a predefined choice.
- comment: String - A comment explaining this [SnippetFindingCuration].
```
> Additional, could their be multiple licenses per snippet ?

license: SpdxExpression so you can take the license array provided by ScanOSS and turn it into expression LICENSE AND LICENSE AND ..

> Regarding the provenance, I think you will run in a wall: ScanOSS for instance does not give any information about the source repository

Actually source code repository is included in the ScanOSS output see https://github.com/oss-review-toolkit/ort/blob/881adaea1d055d5ab63f67abe65b887e2b85d48f/scanner/src/test/assets/scanoss-semver4j-3.1.0-with-snippet.json#L91, was able to translate the ScanOSS Response from [scanoss-semver4j-3.1.0-with-snippet.json](https://github.com/oss-review-toolkit/ort/blob/881adaea1d055d5ab63f67abe65b887e2b85d48f/scanner/src/test/assets/scanoss-semver4j-3.1.0-with-snippet.json) into proposed SnippetFinding

```
{
  location: {
    path: com/vdurmont/semver4j/Range.java,
    startLine: 1,
    endLine: 710
  },
  score: 100,
  detectedSnippet: {
    location: {
      path: src/main/java/com/vdurmont/semver4j/Requirement.java,
      startLine: 1,
      endLine: 710
    },
    purls: [
        "pkg:maven/org.graylog.repackaged/semver4j",
        "pkg:maven/org.graylog2/graylog2-server",
        "pkg:github/graylog2/graylog2-server",
        "pkg:maven/org.graylog/graylog-parent",
        "pkg:maven/org.graylog/graylog-project-parent",
        "pkg:maven/org.graylog.repackaged/os-platform-finder",
        "pkg:maven/org.graylog.telemetry/graylog-telemetry-plugin",
        "pkg:maven/org.graylog.plugins/graylog-plugin-parent",
        "pkg:maven/org.graylog.shaded/kafka09",
        "pkg:maven/org.graylog.shaded/elasticsearch5"
    ],
    vcs: {
       url: "https://github.com/vdurmont/semver4j"
       revision: "b92cd7e4d588747588d1b5fe1f8c664c"
    },
    sourceArtifact: null,
    license: Apache-2.0 AND BSD-2-Clause AND EPL-2.0 AND SSPL
  }
} 
```
Manual curation or generated curation provider for ScanOSS could then resolve purl pkg:maven/org.graylog.repackaged/semver4j into vcs or sourceArtifact info.

Reply:
@nnobelis  Let do a call next week, ping me on Slack on April 12th or later to pick a day & time.

Reply:
> Let do a call next week, ping me on Slack on April 12th or later to pick a day & time.

@nnobelis already scheduled a meeting on April 19 on this topics. Let's just have a *single* meeting *together*.